src/v/storage/disk_log_impl.cc

```diff
@@ -262,31 +262,44 @@ ss::future<> disk_log_impl::garbage_collect_oldest_segments(
 }
 
 ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
-    /*
-     * single segment compaction
-     */
+    // find first not compacted segment
     auto segit = std::find_if(
       _segs.begin(), _segs.end(), [](ss::lw_shared_ptr<segment>& s) {
           return !s->has_appender() && s->is_compacted_segment()
                  && !s->finished_self_compaction();
       });
-    if (segit != _segs.end()) {
+    // loop until we compact segment or reached end of segments set
+    for (; segit != _segs.end(); ++segit) {
         auto seg = *segit;
-        auto f = _stm_manager->ensure_snapshot_exists(
+        if (
+          seg->has_appender() || seg->finished_self_compaction()
+          || !seg->is_compacted_segment()) {
+            continue;
+        }
+
+        co_await _stm_manager->ensure_snapshot_exists(
           seg->offsets().committed_offset);
 
-        return f.then([this, seg, cfg]() {
-            return storage::internal::self_compact_segment(
-                     seg, cfg, _probe, *_readers_cache)
-              .finally([seg] { seg->mark_as_finished_self_compaction(); });
```

Comment:
is `seg->mark_as_finished_self_compaction();` now unnecessary?

Reply:
yes, `mark_as_finished_self_compaction()` is handled inside of `self_compact_segment`